packages/query/src/block-processor.ts

```diff
@@ -0,0 +1,367 @@
+        // position states aren't known by the chain so aren't poulated by
+        // identifyNewAssets
+        // - detect LpNft position opens, and generate all possible position state metadata
+        // - update idb
+        await this.identifyLpNftPositions(Array.from(relevantTx.values()));
+
+        // at this point txinfo can be generated and saved. this will resolve
+        // pending broadcasts, and populate the transaction list.
+        // - calls wasm for each relevant tx
+        // - saves to idb
+        await this.saveTransactionInfo(compactBlock.height, relevantTx);
+      }
+    }
+  }
+
+  private async identifyNewAssets(newNotes: SpendableNoteRecord[]) {
+    for (const n of newNotes) {
+      const assetId = n.note?.value?.assetId;
+      if (!assetId) continue;
+      if (await this.indexedDb.getAssetsMetadata(assetId)) continue;
+
+      let metadata;
+      metadata = await this.querier.shieldedPool.denomMetadata(assetId);
+
+      if (!metadata) {
+        const UNNAMED_ASSET_PREFIX = 'passet';
+        const denom = bech32.encode(UNNAMED_ASSET_PREFIX, bech32.toWords(assetId.inner));
+        metadata = new DenomMetadata({
+          base: denom,
+          denomUnits: [{ aliases: [], denom, exponent: 0 }],
+          display: denom,
+          penumbraAssetId: assetId,
+        });
+      }
+
+      await this.indexedDb.saveAssetsMetadata(metadata);
+    }
+  }
+
+  // Nullifier is published in network when a note is spent or swap is claimed.
+  // - if we possess a matching note or swap record, mark it as spent/claimed
+  // - if we resolve a record sourced from a transaction, and its id is not
+  // recorded, collect and return.
+  private async resolveNullifiers(nullifiers: Nullifier[], height: bigint) {
+    const recordNeedsSourceTxId = new Map<string, SpendableNoteRecord | SwapRecord>();
+
+    // - match nullifiers to idb notes/swaps
+    // - collect any record missing its source
+    // - write to idb.
+    for (const nullifier of nullifiers) {
+      const record =
+        (await this.indexedDb.getNoteByNullifier(nullifier)) ??
+        (await this.indexedDb.getSwapByNullifier(nullifier));
+
+      if (!record) continue;
+
+      // resolve nullifiers for any records of any source type
+      if ('heightSpent' in record) {
+        // spendable note
+        record.heightSpent = height;
+        await this.indexedDb.saveSpendableNote(record);
+      } else if ('heightClaimed' in record) {
+        // swap claimed
+        record.heightClaimed = height;
+        await this.indexedDb.saveSwap(record);
+      }
+
+      // we want to recover ids for transaction sources that don't have an id,
+      // represented by an empty Uint8Array
+      if (
+        record.source?.equals({ source: { case: 'transaction', value: { id: new Uint8Array() } } })
+      )
+        recordNeedsSourceTxId.set(uint8ArrayToHex(nullifier.inner), record);
+    }
```

Comment:
This condition should be removed, we should return spent nullifiers regardless of source